nemo/collections/nlp/models/question_answering/qa_gpt_model.py

```diff
@@ -0,0 +1,485 @@
+                max_length = max(max_length, generated_token_ids[-1].size(1))
+
+            # pad each generated to ensure they are of same length in dim 1, therefore stack-able
+            generated_token_ids = [
+                torch.cat(
+                    [i, torch.ones((1, max_length - i.size(1))).to(i.device) * self.tokenizer.tokenizer.pad_token_id],
+                    axis=-1,
+                )
+                for i in generated_token_ids
+            ]
+            generated_token_ids = torch.cat(generated_token_ids, axis=0)
+            generated_answers = self._get_answers_from_generated_tokens(
+                generated_token_ids, training_mask_end=training_mask_end
+            )
+
+        elif self.cfg.library == 'megatron':
+            raise ValueError("Megatron Generation is not supported by GPTQAModel")
+
+        return generated_answers
+
+    def _get_answers_from_generated_tokens(self, token_ids, training_mask_end=None):
+        answers = []
+        for i in range(token_ids.size(0)):
+            start_point = 0 if training_mask_end is None else training_mask_end[i].item()
+            stop_point = token_ids.size(1)
+
+            for j in range(start_point, stop_point):
+                if token_ids.data[i, j] == self.tokenizer.tokenizer.pad_token_id:
+                    stop_point = j
+                    break
+
+            curr_answer = self.tokenizer.tokenizer.decode(
+                token_ids[i, start_point:stop_point], skip_special_tokens=True
+            ).strip()
+            answers.append(curr_answer)
+
+        return answers
+
+    def _get_raw_scores(self, examples, preds):
+        exact_scores = {}
+        f1_scores = {}
+
+        for example in examples:
+            qas_id = example.qas_id
+            gold_answers = [answer["text"] for answer in example.answers if QAMetrics.normalize_answer(answer["text"])]
+
+            if not gold_answers:
+                # For unanswerable questions, only correct answer is empty string
+                gold_answers = [""]
+
+            if qas_id not in preds:
+                logging.warning(f"Missing prediction for {qas_id}")
+                continue
+
+            pred = preds[qas_id].generated_text
+            exact_scores[qas_id] = max(QAMetrics.get_exact_match(pred, a) for a in gold_answers)
+            f1_scores[qas_id] = max(QAMetrics.get_f1(pred, a) for a in gold_answers)
+
+        return exact_scores, f1_scores
+
+    def _evaluate_predictions(
+        self, examples, all_predictions: Dict[str, str],
+    ):
+        exact, f1 = self._get_raw_scores(examples, all_predictions)
+        total = len(exact)
+
+        return collections.OrderedDict(
+            [
+                ("exact", 100.0 * sum(exact.values()) / total),
+                ("f1", 100.0 * sum(f1.values()) / total),
+                ("total", total),
+            ]
+        )
+
+    def _dump_predictions_to_file(self, output_file, examples, predictions):
```

Comment:
This should go into the metrics class (or a util class if you want one) The reason for this is that dumping predictions has no relation to the model itself

Reply:
What is dumped in the predictions file apart from just the context, question, text would depend on the model though. Ex. start/end pos for BERT extractive QA, ppl for Generative QA. So, formatting the output would still have to be in the model class, so would it make sense to have a separate method in QAMetrics just for dumping a json file?

Reply:
I see - we can keep formatting the output in the model class in this case. Having a dedicated method to dump a json file also seems a little weird since it's only 3 lines. If this is the case, maybe change the name to _format_and_dump_predictions_to_file and also change `output_file` to `output_filename`